src/ledger-ui.c

```diff
@@ -102,11 +118,263 @@ ledger_ui_approve_button(uint32_t mask, uint32_t ctr) {
+              break;
+
+            case HNS_RENEW:
+              strcpy(msg, "RENEW");
+              break;
+
+            case HNS_TRANSFER:
+              strcpy(msg, "TRANSFER");
+              break;
+
+            case HNS_FINALIZE:
+              strcpy(msg, "FINALIZE");
+              break;
+
+            case HNS_REVOKE:
+              strcpy(msg, "REVOKE");
+              break;
+
+            default:
+              THROW(HNS_UNSUPPORTED_COVENANT_TYPE);
+          }
+
+          ledger_ui_update(LEDGER_UI_COVENANT_TYPE, hdr, msg, flags);
+          break;
+        }
+
+        case LEDGER_UI_COVENANT_TYPE: {
+          hns_tx_t *ctx = (hns_tx_t *)g_ledger.ui.ctx;
+          hns_output_t *out = &ctx->curr_output;
+
+          if (out->cov.type == HNS_NONE) {
+            char *hdr = "Value";
+            char *msg = g_ledger.ui.message;
+            volatile uint8_t *flags = g_ledger.ui.flags;
+
+            hex_to_dec(msg, out->val);
+
+            if (!ledger_ui_update(LEDGER_UI_VALUE, hdr, msg, flags))
+              THROW(HNS_CANNOT_UPDATE_UI);
+
+            break;
+          }
+
+          char *hdr = "Name";
+          char *msg = g_ledger.ui.message;
+          volatile uint8_t *flags = g_ledger.ui.flags;
+
+          switch(out->cov.type) {
+            case HNS_OPEN:
+              strcpy(msg, out->cov.items.open.name);
+              break;
+
+            case HNS_BID:
+              strcpy(msg, out->cov.items.bid.name);
+              break;
+
+            case HNS_REVEAL:
+              strcpy(msg, out->cov.items.reveal.name);
+              break;
+
+            case HNS_REDEEM:
+              strcpy(msg, out->cov.items.redeem.name);
+              break;
+
+            case HNS_REGISTER:
+              strcpy(msg, out->cov.items.register_cov.name);
+              break;
+
+            case HNS_UPDATE:
+              strcpy(msg, out->cov.items.update.name);
+              break;
+
+            case HNS_RENEW:
+              strcpy(msg, out->cov.items.renew.name);
+              break;
+
+            case HNS_TRANSFER:
+              strcpy(msg, out->cov.items.transfer.name);
+              break;
+
+            case HNS_FINALIZE:
+              strcpy(msg, out->cov.items.finalize.name);
+              break;
+
+            case HNS_REVOKE:
+              strcpy(msg, out->cov.items.revoke.name);
+              break;
+
+            default:
+              THROW(HNS_UNSUPPORTED_COVENANT_TYPE);
+          }
+
+          ledger_ui_update(LEDGER_UI_NAME, hdr, msg, flags);
+          break;
+        }
+
+        case LEDGER_UI_NAME: {
+          hns_tx_t *ctx = (hns_tx_t *)g_ledger.ui.ctx;
+          hns_output_t *out = &ctx->curr_output;
+
+          if (out->cov.type == HNS_TRANSFER) {
+            char hrp[3];
+            char *hdr = "New Owner";
+            char *msg = g_ledger.ui.message;
+            volatile uint8_t *flags = g_ledger.ui.flags;
+
+            switch(g_ledger.ui.network) {
```

Comment:
Use an array instead of the switch. We can use indices `[0-3]` and use `g_ledger.ui.network >> 1` as the access variable.